### DISCUSSION TITLE
Get list of built pages from revalidate API.

### DISCUSSION THREAD

**Discussion by Gilighost** [vercel/next.js#49123] [Help]
### Summary

Hi all! I am running into an issue with revalidating pages when some of the data used to construct the url is deleted from the CMS.

I have entries in a CMS with fields `id` and `slug`. These fields are used in the paths for the pages with the corresponding entry content. e.g. /content/112u381/title_one

Most of the time I have access to the slug field, so when I want to revalidate the page I can just construct the url and pass it to `res.revalidate`. However, when an entry is deleted from the CMS I no longer have access to the `slug` field, so I cannot construct the URL and call revalidate on it.

Similarly, if the `slug` field is updated, I won't have the old slug on hand to revalidate (I can get it from the CMS, but I can't when it is deleted).

I have been investigating whether I can read the `.next/server/pages` directory from an API route, but it does not include the relevant files.

Related, I am trying to minimize build time and so I do not return the paths from getStaticPaths, just use `fallback: 'blocking'`.

#### Content page `getStaticProps` and `getStaticPaths`: `pages/content/[...slug].js`

```.js
export const getStaticPaths = () => ({ paths: [], fallback: 'blocking' });

export const getStaticProps = async ({
  params,
  preview,
}) => {
  const [contentId, contentSlug] = params?.slug;
  try {
    const content = await getContentById(contentId, preview);

    // If the slug is invalid, redirect to the correct slug
    return content.fields.slug === contentSlug
      ? {
          props: { content },
        }
      : {
          redirect: {
            destination: `/content/${contentId}/${content.fields.slug}`,
            permanent: true,
          },
        };
  } catch (e) {
    console.error(e);
  }
  return {
    notFound: true,
  };
};

```

Ideally, I would be able to retrieve a list of the pages that have been built, and find the relevant path to call revalidate on by content id.

I tested something like this, but it does not work when hosted on Vercel:

```.js
import { glob } from 'glob';
import { promisify } from 'util';

const asyncGlob = promisify(glob);
const outDir = '.next/server/pages';

export default async function handler(req, res) {
  const { globPath } = req.query;
  const searchPath = `${outDir}/${globPath}`;
  const files = await asyncGlob(searchPath, { dot: true });
  const paths = files.map((file) =>
    file.replace(outDir, '').replace('.html', '')
  );
  res.json({ paths });
}

```

Just wondering if anyone has any suggestions. Thanks!

### Additional information

```js
Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 22.3.0: Mon Jan 30 20:42:11 PST 2023; root:xnu-8792.81.3~2/RELEASE_X86_64
    Binaries:
      Node: 18.13.0
      npm: 8.19.3
      Yarn: N/A
      pnpm: 8.0.0
    Relevant packages:
      next: 13.3.4
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
```

**Comment by guilherme-marin** [comment]
Hello @Gilighost did you manage to find any alternative for this? I'm facing almost the same problem, I'd like to revalidate an entire route with /content/[...slug] since I'm using fallback blocking as well, so I don't know the paths that were created. 

I wish there was a way to reset an entire path, or preferably know which pages were already generated related to the my content. Since my application's filters already generate static pages, there's no way to know which ones have already been generated AFAIK.
